server/mailet/dkim/src/main/java/org/apache/james/jdkim/mailets/MimeMessageHeaders.java

```diff
@@ -19,48 +19,49 @@
 
 package org.apache.james.jdkim.mailets;
 
-import java.util.Enumeration;
-import java.util.HashMap;
-import java.util.LinkedList;
 import java.util.List;
-import java.util.Map;
 
 import javax.mail.MessagingException;
 import javax.mail.internet.MimeMessage;
 
+import com.github.fge.lambdas.Throwing;
+import com.github.steveash.guavate.Guavate;
 import com.google.common.collect.ImmutableList;
-import com.google.common.collect.ImmutableMap;
+import com.google.common.collect.ImmutableListMultimap;
+import org.apache.commons.lang3.tuple.Pair;
 import org.apache.james.jdkim.api.Headers;
+import org.apache.james.util.StreamUtils;
 
 /**
  * An adapter to let DKIMSigner read headers from MimeMessage
  */
 final class MimeMessageHeaders implements Headers {
 
-    private final Map<String, List<String>> headers;
+    private final ImmutableListMultimap<String, String> headers;
     private final List<String> fields;
 
     public MimeMessageHeaders(MimeMessage message) throws MessagingException {
-        Map<String, List<String>> headersBuilding = new HashMap<>();
-        ImmutableList.Builder<String> fieldsBuilder = ImmutableList.builder();
-        for (Enumeration<String> e = message.getAllHeaderLines(); e.hasMoreElements();) {
-            String head = e.nextElement();
-            int p = head.indexOf(':');
-            if (p <= 0) {
-                throw new MessagingException("Bad header line: " + head);
-            }
-            String headerName = head.substring(0, p).trim();
-            String headerNameLC = headerName.toLowerCase();
-            fieldsBuilder.add(headerName);
-            List<String> strings = headersBuilding.get(headerNameLC);
-            if (strings == null) {
-                strings = new LinkedList<>();
-                headersBuilding.put(headerNameLC, strings);
-            }
-            strings.add(head);
-        }
-        headers = ImmutableMap.copyOf(headersBuilding);
-        fields = fieldsBuilder.build();
+        ImmutableList<Pair<String, String>> headsAndLines = StreamUtils
```

Comment:
IMO 

```
ImmutableList<Header> headers = StreamUtils.ofEnum(message.getAllHeaders())
    .collect(Guavate.toImmutableList());
```

sounds waaaay better than doing the parsing work ourselves...

Not to mention that we now have stronger types...